service/src/main/java/org/cbioportal/service/impl/GenesetDataServiceImpl.java

```diff
@@ -0,0 +1,108 @@
+import org.cbioportal.persistence.SampleListRepository;
+import org.cbioportal.service.GenesetDataService;
+import org.cbioportal.service.GeneticProfileService;
+import org.cbioportal.service.SampleService;
+import org.cbioportal.service.exception.GeneticProfileNotFoundException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class GenesetDataServiceImpl implements GenesetDataService {
+
+    @Autowired
+    private GeneticDataRepository geneticDataRepository;
+    @Autowired
+    private SampleService sampleService;
+    @Autowired
+    private GeneticProfileService geneticProfileService;
+    @Autowired
+    private SampleListRepository sampleListRepository;
+
+    @Override
+	public List<GenesetData> fetchGenesetData(String geneticProfileId, List<String> sampleIds, List<String> genesetIds)
+			throws GeneticProfileNotFoundException {
+
+        List<GenesetData> genesetDataList = new ArrayList<>();
+
+        String commaSeparatedSampleIdsOfGeneticProfile = geneticDataRepository
+            .getCommaSeparatedSampleIdsOfGeneticProfile(geneticProfileId);
+        if (commaSeparatedSampleIdsOfGeneticProfile == null) {
+        	//no data, return empty list:
+            return genesetDataList;
+        }
+        List<Integer> internalSampleIds = Arrays.stream(commaSeparatedSampleIdsOfGeneticProfile.split(","))
+            .mapToInt(Integer::parseInt).boxed().collect(Collectors.toList());
+
+        List<Sample> samples;
+        if (sampleIds == null) {
+            samples = sampleService.getSamplesByInternalIds(internalSampleIds);
+        } else {
+            GeneticProfile geneticProfile = geneticProfileService.getGeneticProfile(geneticProfileId);
+            List<String> studyIds = new ArrayList<>();
+            sampleIds.forEach(s -> studyIds.add(geneticProfile.getCancerStudyIdentifier()));
+            samples = sampleService.fetchSamples(studyIds, sampleIds, "ID");
+        }
+
+        List<GenesetAlteration> genesetAlterations = geneticDataRepository.getGenesetAlterations(geneticProfileId,
+        		genesetIds, "SUMMARY");
+        
+        for (Sample sample : samples) {
+            int indexOfSampleId = internalSampleIds.indexOf(sample.getInternalId());
+            if (indexOfSampleId != -1) {
+                for (GenesetAlteration genesetAlteration : genesetAlterations) {
+                    GenesetData genesetData = new GenesetData();
+                    genesetData.setGeneticProfileId(geneticProfileId);
+                    genesetData.setSampleId(sample.getStableId());
+                    genesetData.setGenesetId(genesetAlteration.getGenesetId());
+                    genesetData.setValue(genesetAlteration.getSplitValues()[indexOfSampleId]); //TODO introduce this "splitValues cache" in genetic data API as well...since .split(",") turned out to be bottleneck here...service went from 4s to 165ms after this improvement!
```

Comment:
@pieterlukasse that's right! I refactored the method a few times, didn't realize that I was calling the `split` on the same alteration object for every sample, nice catch!